draft-ietf-httpbis-cache-digest.md

```diff
@@ -161,85 +159,147 @@ When generating CACHE_DIGEST, a client MUST NOT include cached responses whose U
+6. Let `fingerprint` be the return value of {{fingerprint}} with `key`, `N` and `f` as inputs.
+7. Let `h2` be the return value of {{hash}} with `fingerprint` and `N` as inputs, XORed with `h1`.
+8. Let `h` be either `h1` or `h2`, picked in random.
+9. Let `position_start` be 40 + `h` * `f`.
+10. Let `position_end` be `position_start` + `f` * `b`.
+11. While `position_start` < `position_end`:
+    1. Let `bits` be `f` bits from `digest_value` starting at `position_start`.
+    2. If `bits` is all zeros, set `bits` to `fingerprint` and terminate these steps.
+    3. Add `f` to `position_start`.
+12. Substract `f` from `position_start`.
+13. Let `fingerprint` be the `f` bits starting at `position_start`.
+14. Let `h1` be `h`
+15. Substract 1 from `maxcount`.
+16. If `maxcount` is zero, return an error.
+17. Go to step 7.
+
+
+### Removing a URL to the Digest-Value {#removing}
 
+Given the following inputs:
 
-### Computing a Hash Value {#hash}
+* `URL` a string corresponding to the Effective Request URI ({{RFC7230}}, Section 5.5) of a cached response {{RFC7234}}
+* `ETag` a string corresponding to the entity-tag {{RFC7232}} if a cached response {{RFC7234}} (if the ETag is available; otherwise, null);
+* `digest-value`
+
+1. Let `f` be the value of the first byte of `digest-value`.
+3. Let `N` be the value of the second to fifth bytes of `digest-value` in big endian form.
+4. Let `key` be the return value of {{key}} with `URL` and `ETag` as inputs.
+5. Let `h1` be the return value of {{hash}} with `key` and `N` as inputs.
+6. Let `fingerprint` be the return value of {{fingerprint}} with `key`, `N` and `f` as inputs.
+7. Let `h2` be the return value of {{hash}} with `fingerprint` and `N` as inputs, XORed with `h1`.
+8. Let `h` be `h1`.
+9. Let `position_start` be 40 + `h` * `f`.
+10. Let `position_end` be `position_start` + `f` * `b`.
+11. While `position_start` < `position_end`:
+    1. Let `bits` be `f` bits from `digest_value` starting at `position_start`.
+    2. If `bits` is `fingerprint`, set `bits` to all zeros and terminate these steps.
+    3. Add `f` to `position_start`.
+12. If `h` is not `h2`, set `h` to `h2` and return to step 9.
+
+### Computing a fingerprint value {#fingerprint}
 
-Given:
+Given the following inputs:
 
-* `URL`, an array of characters
-* `ETag`, an array of characters
-* `validators`, a boolean
+* `key`, an array of characters
 * `N`, an integer
-* `P`, an integer
+* `f`, an integer indicating the number of output bits
 
-`hash-value` can be computed using the following algorithm:
+1. Let `hash-value` be the return value of {{hash}} with `key` and `N` as inputs.
+2. Let `h` be the number of bits in `hash-value`
+3. Let `fingerprint-value` be 0
+4. While `fingerprint-value` is 0 and `h` > `f`:
+    4.1. Let `fingerprint-value` be the `f` least significant bits of `hash-value`.
+    4.2. Let `hash-value` be the the `h`-`f` most significant bits of `hash-value`.
+    4.3. `h` -= `f`
```

Comment:
This code feels inconsistent with the writing style used throughout. Would suggest:
> Substract `f` from `h`.